### PR TITLE
Reintroduce k8s client fallback to cache lookups

### PR DESCRIPTION
**Pull request by hamersaw**
## Tracking issue
fixes https://github.com/flyteorg/flyte/issues/4730

## Why are the changes needed?
[This PR](https://github.com/flyteorg/flyte/pull/4199) updated k8s and removed the logic to fallback to the client `Reader` when the cache `Reader` was a miss. In high traffic scenarios the k8s informer can not populate the cache fast enough, so the fallback API call is required to ensure correctness.

## What changes were proposed in this pull request?
Creating the k8s client with a fallback to API call when an object does not exist in the informer cache.

## How was this patch tested?
Faked an object `NotFound` error because of cache miss.

### Setup process
Updated k8s [client](https://github.com/flyteorg/flyte/blob/0c8dc61833683c7d2482156527c8f45616ee4a20/flytestdlib/otelutils/k8s.go#L43-L47) and [cache](https://github.com/flyteorg/flyte/blob/master/flytestdlib/otelutils/k8s.go#L21-L25) through the otel wrapper to fail retrievals if the object was created less than 3 seconds ago. Example code:
```
    func (c *K8sCacheWrapper) Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
        ctx, span := NewSpan(ctx, K8sClientTracer, fmt.Sprintf("%s.Cache/Get", k8sSpanPathPrefix))
        defer span.End()

        err := c.Cache.Get(ctx, key, obj, opts...)
        if time.Since(obj.GetCreationTimestamp().Time) < (time.Second * 3) {
            obj = nil
            return k8serrors.NewNotFound(schema.GroupResource{Group: "foo", Resource: "bar"}, key.Name)
        }

        return err
        //return c.Cache.Get(ctx, key, obj, opts...)
    }
```

Then validated that this causes Pods to be stuck in `Terminating` state with finalizers still intact:
```
    hamersaw@ragnarok:~$ kubectl -n flytesnacks-development get pods
    NAME                        READY   STATUS        RESTARTS   AGE
    f5329181f1364487b8a9-n0-0   0/1     Terminating   0          95s
    f5329181f1364487b8a9-n0-1   0/1     Terminating   0          86s
    f5329181f1364487b8a9-n0-2   0/1     Terminating   0          76s
```
with log
```
{"json":{"exec_id":"f5329181f1364487b8a9","node":"n0","ns":"flytesnacks-development","res_ver":"11272","routine":"worker-0","src":"plugin_manager.go:269","tasktype":"python-task","wf":"flytesnacks:development:hello-world.my_wf_no_outputs_5"},"level":"warning","msg":"Failed to find the Resource with name: flytesnacks-development/f5329181f1364487b8a9-n0-0. Error: bar.foo \"f5329181f1364487b8a9-n0-0\" not found","ts":"2024-01-16T14:28:07-06:00"}
```
and UI message
`[3/3] currentAttempt done. Last Error: SYSTEM::resource not found, name [flytesnacks-development/f5329181f1364487b8a9-n0-3]. reason: bar.foo "f5329181f1364487b8a9-n0-3" not found`

### Screenshots
_NA_

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
_NA_

## Docs link
_NA_

### CODE REVIEW

**Comment by honnix** [comment]
Just to confirm my understanding. Does it mean [this comment](https://github.com/flyteorg/flyte/blob/0c8dc61833683c7d2482156527c8f45616ee4a20/flytepropeller/pkg/controller/nodes/task/k8s/plugin_manager.go#L264) is not correct without this PR?

**Comment by hamersaw** [comment]
> Just to confirm my understanding. Does it mean [this comment](https://github.com/flyteorg/flyte/blob/0c8dc61833683c7d2482156527c8f45616ee4a20/flytepropeller/pkg/controller/nodes/task/k8s/plugin_manager.go#L264) is not correct without this PR?

@honnix correct, previous to [this merge](https://github.com/flyteorg/flyte/pull/4199) it was and ti will be again after we get this PR merged. Right now, we only look at the informer cache and do not fall back to the api reader.

File: flytepropeller/pkg/controller/executors/kube.go [hunk 1]
```diff
@@ -23,26 +24,66 @@ type Client interface {
...
+
+	return
 }
 
-// NewFallbackClientBuilder Creates a new k8s client that uses the cached client for reads and falls back to making API
-// calls if it failed. Write calls will always go to raw client directly.
-func NewFallbackClientBuilder(scope promutils.Scope) *FallbackClientBuilder {
-	return &FallbackClientBuilder{
-		scope: scope,
+func (c fallbackClientReader) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) (err error) {
```
Comment (kumare3) [inline flytepropeller/pkg/controller/executors/kube.go:79]: I remember this code

**Review by kumare3 — APPROVED** [review]
Lgtm

**Comment by honnix** [comment]
> > Just to confirm my understanding. Does it mean [this comment](https://github.com/flyteorg/flyte/blob/0c8dc61833683c7d2482156527c8f45616ee4a20/flytepropeller/pkg/controller/nodes/task/k8s/plugin_manager.go#L264) is not correct without this PR?
> 
> @honnix correct, previous to [this merge](https://github.com/flyteorg/flyte/pull/4199) it was and ti will be again after we get this PR merged. Right now, we only look at the informer cache and do not fall back to the api reader.

This comment successfully tricked me not to look further into client details, lol.

**Pull request merged**